onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,702 @@
+                21, // Constant count
+                &constants,
+                0 // offset
+            );
+
+            commandList->Dispatch(dispatchSizeX, 1, 1);
+        }
+
+        commandList->ResourceBarrier(2, uav_barriers);
+    }
+
+    void BluesteinZChirp(
+        ID3D12Resource* /*inputResource*/,
+        ID3D12Resource* /*outputResource*/,
+        ID3D12GraphicsCommandList* /*commandList*/)
+    {
+        // Not implemented
+        throw;
+    }
+};
+
+struct DFTShapeInferrer : public WRL::Base<IMLOperatorShapeInferrer>
+{
+    STDMETHOD(InferOutputShapes)(IMLOperatorShapeInferenceContext* context) noexcept
+    {
+        try
+        {
+            int64_t axis;
+            ORT_THROW_IF_FAILED(context->GetAttribute("axis", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&axis)));
+            int64_t inverse_i;
+            ORT_THROW_IF_FAILED(context->GetAttribute("inverse", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&inverse_i)));
+            int64_t onesided_i;
+            ORT_THROW_IF_FAILED(context->GetAttribute("onesided", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&onesided_i)));
+            bool is_onesided = static_cast<bool>(onesided_i);
+            bool inverse = static_cast<bool>(inverse_i);
+
+            if (inverse && is_onesided) {
+                throw new std::exception("is_onesided and inverse attributes cannot be enabled at the same time");
+            }
+
+            uint32_t rank;
+            ORT_THROW_IF_FAILED(context->GetInputTensorDimensionCount(0, &rank));
+            if (rank == 0) {
+                // If no shape is available for the input, skip shape inference...
+                throw;
+            }
+
+            if (context->IsInputValid(1))
+            {
+                // If dft_length is specified, then we should honor the shape.
+                // Set the output dimension to match the dft_length on the axis.
+                // If onesided this will be adjusted later on... this is not supported as of now
+                // do nothng... for some reason unset optional inputs seem to crash..
+            }
+
+            // In general the output shape will match the input shape exactly
+            // So initialize the output shape with the input shape
+            std::vector<uint32_t> inputDims(rank);
+            ORT_THROW_IF_FAILED(context->GetInputTensorShape(0, rank, inputDims.data()));
+            auto outputDims = inputDims;
+            outputDims.back() = 2;
+
+            if (!(-(int)rank <= (int)axis && axis < rank)) {
+                std::stringstream ss;
+                ss << "axis attribute value " << axis << " is invalid for a tensor of rank " << rank;
+                auto ss_str = ss.str();
+                throw new std::exception(ss_str.c_str());
+            }
+
+            auto axis_idx = (axis >= 0 ? axis : axis + rank);
```

Comment:
There is an existing helper function [OperatorHelper::HandleNegativeAxis](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp#L16), which can be used here.

Reply:
updated, thanks for the tip!